README

Libgwyfile is a small MIT-licensed embeddable library that provides functions
for reading and writing of Gwyddion GWY files. It contains functions for both
the generic GWY file format and specific data structures actually used by
Gwyddion to represent scanning probe microscopy data.

Website: http://libgwyfile.sourceforge.net/
Author:  Yeti <yeti@gwyddion.net>

-----------------------------------------------------------------------------

0. REQUIREMENTS

An ISO C compiler providing a few C99 features, in particular:
- the inline keyword
- stdbool.h
- stdint.h

Libgwyfile has no external dependencies.

-----------------------------------------------------------------------------

1. BUILDING AS A NORMAL (SHARED/STATIC) LIBRARY ON UNIX

This may be a good start even if you intend to embed the library to ensure it
works fine on your system.

Libgwyfile uses GNU autotools so run

    ./configure [--prefix=...]
    make
    make install

as usual to build it.

Documentation.  If you have doxygen you can run

    make doc

to HTML build API reference documentation in lib/doc/.  Read it on-line
otherwise.

Tests.  If you have valgrind you can run

    make check

to run the tests.  It runs each test program once normally and once under
valgrind.  Of course, the tests can be run also without valgrind.  However, as
most of them specifically excersise ownership transfer and memory handling they
are much more meaningful under valgrind.

You might be able to run tests normally but not under valgrind.  This can
typically happen when you are cross-compiling and have a binary interpreter for
the host system (such as wine), but not valgrind.  In such case either pass
--without-valgrind to configure or use

    make check valgrind_tests=

-----------------------------------------------------------------------------

2. WIN32 CROSS-COMPILATION ON UNIX

Usually, you would just embed the gwyfile.[ch] into your Win32 project.
Anyway, cross-compilation on Unix is possible and if you have Wine you can run
the examples and the tests (and they should pass).

Run configure as

    ./configure --prefix=<PREFIX> --host=<HOST> --build=<BUILD>

where <HOST> is either i686-w64-mingw32 or x86_64-w64-mingw32 depending on the
target architecture (32bit or 64bit) and <BUILD> is whatever the config.guess
script from Automake prints on your system.  It is typically something like
x86_64-unknown-linux-gnu.

In principle you do not need to pass the --build option.  However, if you have
Wine installed to run the tests and examples and the correspondign executable
format handler is registered, configure may fall under the impression you are
doing a native compilation (as opposed to cross-compiling).  Specifying the
--build option ensures the build is seen as a cross-compilation.

The final DLLs are produced upon installation (make install) so you have to
do that to get them.  Specify any temporary <PREFIX> of your choice as it
makes little sense to do a system-wide installation in this case.

-----------------------------------------------------------------------------

3. BUILDING AS A NORMAL LIBRARY WITH MSVC

MSVC 2015 project files are provided in subdirectory msvc2015.  The SLN file
is located in msvc2015/libgwyfile.  The files use relative paths to the
source and should work from the location where they distributed in the
archive.

-----------------------------------------------------------------------------

4. EMBEDDING THE LIBRARY

The entire library consists of two C files in lib:

    gwyfile.h (the header)
    gwyfile.c (the implementation)

Just copy them to your program.

See the top of gwyfile.c for preprocessor defines that you may want to adjust
to define the system configuration.  Most importantly, ensure the library is
built for the correct byte order (little/big endian).  Alternatively, you can
provide a config.h that defines the system configuration.

If your C compiler does not support C99 features such as the bool type or
inline keyword, see the top of gwyfile.c for some possible remedies.

-----------------------------------------------------------------------------

5. USAGE EXAMPLES

See the files in examples subdirectory:
- readgwy (reads Gwyddion GWY files and print some information about the data)
- writegwy (writes a simple Gwyddion GWY file with some channel and graph)
- readgeneric (reads generic GWY files and prints their structure)
- writegeneric (writes a simple generic GWY file)
- checkgeneric (checks GWY files for problems that go against the specs but
  do not prevent libgwyfile from loading the file)



NEWS
libgwyfile-1.6 (2023-04-24)
- Fixed mishandling of GwyLawn "segments" data which should contain twice as
  many integers than gwyfile 1.5 expected.

libgwyfile-1.5 (2022-02-11)
- New functions for GwyLawn curve map data type.
- New functions for GwySelectionProjective selection type.
- New flag for checking Gwyddion-specific data objects.
- Examples using gwyfile to send messages over network were improved to handle
  various failures in realistic situations better.
- Fixed gwyfile_object_container_enumerate_xyz() trying to access Surfaces as
  DataFields.

libgwyfile-1.4 (2020-06-08)
- New functions for item and object copying.
- A simple example using gwyfile to send messages over network was added.
- A specific error code for truncated buffers was added.
- The Win32 library symbol file gwyfile.def was actually updated.
- Declaration of function gwyfile_object_data_size() was restored.
- Vim syntax file was added.

libgwyfile-1.3 (2020-06-02)
- New functions for reading from and writing to memory buffers.
- New functions for GwySelectionCross selection type.
- New function for accessing n-th item of an object.
- Win32 compilation error in gwyfile_set_error_errno() was fixed.
- Byte swapping when writing files was fixed for big endian machines.
- Documentation build on Debian was fixed.

libgwyfile-1.2 (2016-04-27)
- New functions for handling GwySurface and GwySelectionPath data objects were
  added.
- New function for enumeration of XYZ data in Gwyddion files was added.
- Example program readgwy.c was updated to include the new data types.
- Installing the library (as opposed to embedding it) works meaningfully, i.e.
  it actually installs the library files and headers now.  A pkg-config file
  is also installed now.

libgwyfile-1.1 (2016-01-12)
- gwyfile_item_get_int64() and gwyfile_item_new_internal_int64() were fixed
  to actually work with 64bit values, not just 32bit.
- New function gwyfile_check_object() can check for violation of the GWY file
  format specifications that are not physically enforced during normal
  libgwyfile usage, and for things that are not violations but should be
  avoided anyway.
- New struct GwyfileErrorList was added for gwyfile_check_object() reports.
- A few error messages are more specific or detailed now.
- New example, checkgeneric, was added as a simple demonstration of the
  checking functions.
- Running tests under valgrind can be disabled by --without-valgrind configure
  option.

libgwyfile-1.0 (2015-07-13)
- Special item names for data ownership taking (ending with "(take)") now
  actually work instead of returning NULL.
- GwyDataLine data extractor was fixed to extract "res", not "xres".
- GwySelectionAxis constructor actually returns GwySelectionAxis now, not
  GwySelectionLattice.
- Non-standard use of va_arg() was fixed, making convenience constructors and
  extractors work when compiled with MSVC.
- The maximum object/item nesting depth has been limited to 200 in file reading
  to avoid stack overflows.
- It is safe to call gwyfile_error_clear() with the pointer to an unset error.
- New functions gwyfile_write_wfile() and gwyfile_read_wfile() were added for
  Win32 to permit using wide-character file names.
- A workaround for broken vsnprintf() on Win32 was added to error reporting so
  detailed errors have now meaningful domains, codes and messages on Win32.
- Examples were tweaked to compile on Win32 cleanly.
- Misdetection of Win32 environment by configure was fixed.
- A gwyfile.def listing exported symbols was added to the distribution for
  Win32 compilation.
- MSVC 2015 project files were added to the distribution in the msvc2015
  subdirectory.
- Examples are linked as static.

libgwyfile-0.9 (2015-03-31)
- Many more functions for construction of Gwyddion data objects.
- The special item names specifying memory handling have changed from e.g.
  "data-copy" to "data(copy)".  And they actually work now.
- New functions for extracting information from Gwyddion data objects:
  gwyfile_object_datafield_get(), gwyfile_object_dataline_get(), etc.
- New high-level functions for enumeration of channels, volume data, etc. in
  Gwyddion files.
- New functions for obtaining the number of items in an object and the list of
  their names.
- Construction of item lists, either using gwyfile_object_newv() or when
  reading files, is no longer quadratic-time on the number of items.
- Array setter functions correctly update the array length to the new length.
- gwyfile_item_release_object() unsets the object owner as it should.
- A confinement error is reported when the end of file is hit unexpectedly in
  otherwise valid data (previous version reported a system error with zero
  errno).
- gwyfile_fread() actually uses the max_size argument to constrain the size
  of the top-level object.
- gwyfile_read_file() determines the file size first for improved overrun
  checks.
- Miscellaneous documentation improvements.

libgwyfile-0.8 (2014-12-14)
- First public release.


examples/writegeneric.c
/*
 * $Id: writegeneric.c 343 2020-07-10 09:36:51Z yeti-dn $
 *
 * This example shows how to write a simple generic GWY file "test.gwy"
 * containing all kinds of basic datatypes.
 *
 * I, the copyright holder of this work, release this work into the public
 * domain.  This applies worldwide.  In some countries this may not be legally
 * possible; if so: I grant anyone the right to use this work for any purpose,
 * without any conditions, unless such conditions are required by law.
 */
#include <stdlib.h>
#include <math.h>
#include <errno.h>
#include <string.h>
#include "gwyfile.h"

/* Array length. */
#define AL(x) (sizeof(x)/sizeof((x)[0]))

enum {
    OBJECTS_DATA_LEN = 2,
    STRINGS_DATA_LEN = 3,
};

static char char_data[21];
static int32_t int32_data[6];
static int64_t int64_data[15];
static double double_data[13];

int
main(void)
{
    GwyfileObject *gwyf, **objects_data;
    GwyfileError *error = NULL;
    char **strings_data;
    unsigned int i;

    /* Build the file structure.  We use the _copy functions here so the
     * arrays will not be consumed.  */
    gwyf = gwyfile_object_new("MyStruct",
                              gwyfile_item_new_bool("bool", true),
                              gwyfile_item_new_char("char", 'X'),
                              gwyfile_item_new_int32("int32", 42),
                              gwyfile_item_new_int64("int64", 1234567890ULL),
                              gwyfile_item_new_double("double", 3.14),
                              gwyfile_item_new_string_copy("string", "Shark"),
                              gwyfile_item_new_char_array_copy("chars", char_data, AL(char_data)),
                              gwyfile_item_new_int32_array_copy("int32s", int32_data, AL(int32_data)),
                              gwyfile_item_new_int64_array_copy("int64s", int64_data, AL(int64_data)),
                              gwyfile_item_new_double_array_copy("doubles", double_data, AL(double_data)),
                              NULL);

    /* Create a string array.  This is a bit more involved.  Note the array
     * strings_data[] will be also consumed. */
    strings_data = (char**)malloc(sizeof(char*)*STRINGS_DATA_LEN);
    for (i = 0; i < STRINGS_DATA_LEN; i++) {
        strings_data[i] = (char*)malloc(8);
        snprintf(strings_data[i], 8, "string%u", i);
    }
    gwyfile_object_add(gwyf, gwyfile_item_new_string_array("strings", strings_data, STRINGS_DATA_LEN));

    /* Create an object array.  This is a bit more involved.  Note the array
     * objects_data[] will be also consumed. */
    objects_data = (GwyfileObject**)malloc(sizeof(GwyfileObject*)*OBJECTS_DATA_LEN);
    for (i = 0; i < OBJECTS_DATA_LEN; i++)
        objects_data[i] = gwyfile_object_new("MySubStruct", gwyfile_item_new_int32("i", i), NULL);
    gwyfile_object_add(gwyf, gwyfile_item_new_object_array("objects", objects_data, OBJECTS_DATA_LEN));

    /* Write the top-level object to a file. */
    if (!gwyfile_write_file(gwyf, "test.gwy", &error)) {
        fprintf(stderr, "Cannot write test.gwy: %s\n", error->message);
        gwyfile_error_clear(&error);
        exit(EXIT_FAILURE);
    }

    /* Clean-up.  With the default ownership transfer rules, this call nicely
     * frees all the created objects and items. */
    gwyfile_object_free(gwyf);

    return 0;
}

/* vim: set cin et ts=4 sw=4 columns=120 tw=119 cino=>1s,e0,n0,f0,{0,}0,^0,\:1s,=0,g1s,h0,t0,+1s,c3,(0,u0 : */
